README
Evil 1.7.4

The Evil library tried to port some convenient Unix functions
to the Windows (XP or CE) platform. They are planned to be used
in the Enlightenment Fundations Libaries (EFL).

To get information on how to build that library and the EFL on
Windows XP, see

http://trac.enlightenment.org/e/wiki/EFLWindowsXP

To get information on how to build that library and the EFL on
on Windows CE, see


http://trac.enlightenment.org/e/wiki/EFLWindowsCE

To get the description of the API, behavior, etc.. run 'make doc'
in the top level directory, then open in your favorite browser
named Firefox the file index.html located in the subdirectory
doc/html/.

Enjoy !


NEWS
Evil NEWS - User visible changes.

* Evil 1.7.9:

** fix compilation for Windows >= Vista and gcc 4.8.1.

* Evil 1.7.8:

** Bumping version to match latest release, no changes made.

* Evil 1.7.7:

** Bumping version to match latest release, no changes made.

* Evil 1.7.6:

** Bumping version to match latest release, no changes made.

* Evil 1.7.5:

** Bumping version to match latest release, no changes made.

* Evil 1.7.4:

** Bumping version to match latest release, no changes made.

* Evil 1.7.2:

** Bumping version to match latest release, no changes made.

* Evil 1.7:

** Released August 30th, 2012, by Carsten Haitzler
** Correctly display messages with evil_format_message()

* Evil 1.1:

** Released April 26th, 2012, by Vincent Torri
** Add evil_path_is_absolute() API
** Add POSIX printf() family functions
** Add S_ISLNK macro
** Add strcasestr() API
** Do not declare and define localtime_r() if it's already defined
** Fix nl_langinfo(RADIXCHAR)

* Evil 1.0:

** Released December 2nd, 2011, by Vincent Torri


src/lib/Evil.h
#ifndef __EVIL_H__
#define __EVIL_H__

/**
 * @mainpage Evil
 * @image html  e_big.png
 * @author Vincent Torri
 * @version 1.7.5
 * @date 2008-2013
 *
 * @section intro_sec Introduction
 *
 * The Evil library is an evil library that ports some evil Unix
 * functions to the Windows (XP or above, or Mobile) platform. The
 * evilness is so huge that the most of the functions are not POSIX or
 * BSD compliant.
 *
 * These functions are intended to be used in the Enlightenment
 * Foundation Libraries only and can be compiled only on Windows,
 * using MSYS/MinGW on Windows, and cross-compilation on Unix. This
 * library is minimal in the sense that only the functions needed to
 * compile the EFL are available. The purpose of this library is NOT
 * to have a full POSIX emulation et it is NOT a replacement of
 * cygwin. To compare the size of the DLL themselves, Evil is around
 * 33 KB and cygwin DLL is around 800 KB.
 *
 * @section acknowledgments_sec Acknowledgments
 *
 * This library has receive some from people interested in the EFL or
 * not. Among them, evil thanks to Lars Munch, Raoul Hecky, Nicolas
 * Aguirre, Tor Lillqvist, Lance Fetters, Vincent Richomme, Paul
 * Vixie, Daniel Stenberg, who helped the author of the library in
 * different fields (code and tests).
 *
 * @section license_sec license
 *
 * The Evil library is distributes under a modified BSD license. See
 * the files COPYING and COPYING-PLAIN in the top level directory for
 * the full license text.
 *
 * @section reference_sec Reference API
 *
 * Use the horizontal menu above to navigate into the reference API
 */

/**
 * @file Evil.h
 * @brief The file that provides miscellaneous functions ported from Unix.
 * @defgroup Evil Miscellaneous functions ported from Unix.
 *
 * This header provides miscellaneous functions that exist on Unix
 * but not on Windows platform. They try to follow the conformance of
 * the Unix versions.
 */

/**
 * @cond LOCAL
 */

#ifdef __cplusplus
extern "C" {
#endif


#ifndef WIN32_LEAN_AND_MEAN
# define WIN32_LEAN_AND_MEAN
#endif
#include <windows.h>
#undef WIN32_LEAN_AND_MEAN

#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <limits.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <math.h>
#include <direct.h>


#ifdef _MSC_VER

# include <io.h>

# define F_OK 0  /* Check for file existence */
# define X_OK 1  /* MS access() doesn't check for execute permission. */
# define W_OK 2  /* Check for write permission */
# define R_OK 4  /* Check for read permission */

typedef DWORD          pid_t;
typedef unsigned short mode_t;

typedef unsigned short uint16_t;
typedef unsigned int uint32_t;
typedef signed int int32_t;
typedef __int64 int64_t;
typedef unsigned __int64 uint64_t;
typedef SSIZE_T ssize_t;

# define strdup(s) _strdup(s)
# define unlink(filename) _unlink(filename)
# define fileno(f) _fileno(f)
# define fdopen(fd,m) _fdopen((fd),(m))
# define access(p,m) _access((p),(m))
# define hypot(x,y) _hypot((x),(y))
# define tzset _tzset

#endif /* _MSC_VER */

#ifdef _WIN32_WCE
# ifndef offsetof
#  define offsetof(type, ident) ((size_t)&(((type*)0)->ident))
# endif
#endif

typedef unsigned long  uid_t;
typedef unsigned long  gid_t;


#include "evil_macro.h"
#include "evil_fcntl.h"
#include "evil_inet.h"
#include "evil_langinfo.h"
#include "evil_libgen.h"
#include "evil_main.h"
#include "evil_print.h"
#include "evil_stdlib.h"
#include "evil_stdio.h"
#include "evil_string.h"
#include "evil_time.h"
#include "evil_unistd.h"
#include "evil_util.h"
#include "evil_macro_pop.h"


#if (defined(_WIN32) && !defined(_UWIN) && !defined(__CYGWIN__))
# if defined(_MSC_VER) || defined(__MINGW32__)

# ifdef S_ISDIR
#  undef S_ISDIR
# endif
# ifdef S_ISREG
#  undef S_ISREG
# endif
# define S_ISDIR(m) (((m) & _S_IFMT) == _S_IFDIR)
# define S_ISREG(m) (((m) & _S_IFMT) == _S_IFREG)

# define S_ISLNK(m) 0

# define S_IRUSR _S_IRUSR
# define S_IWUSR _S_IWUSR
# define S_IXUSR _S_IXUSR
# define S_IRGRP S_IRUSR
# define S_IROTH S_IRUSR
# define S_IWGRP S_IWUSR
# define S_IWOTH S_IWUSR
# define S_IXGRP S_IXUSR
# define S_IXOTH S_IXUSR

# define _S_IRWXU (_S_IREAD | _S_IWRITE | _S_IEXEC)
# define _S_IXUSR _S_IEXEC
# define _S_IWUSR _S_IWRITE
# define _S_IRUSR _S_IREAD

# define S_IRWXO _S_IRWXU
# define S_IRWXG _S_IRWXU

#  define mkdir(p,m) _mkdir(p)
  /*
#  define close(fd) _close(fd)
#  define read(fd,buffer,count) _read((fd),(buffer),(count))
#  define write(fd,buffer,count) _write((fd),(buffer),(count))
#  define unlink(filename) _unlink((filename))
#  define lstat(f,s) _stat((f),(s))
  */

# endif
#endif


#ifdef __cplusplus
}
#endif

/**
 * @endcond
 */

#endif /* __EVIL_H__ */


src/lib/evil_inet.h

/* Copyright (C) 2005 by Daniel Stenberg
 *
 * Permission to use, copy, modify, and distribute this
 * software and its documentation for any purpose and without
 * fee is hereby granted, provided that the above copyright
 * notice appear in all copies and that both that copyright
 * notice and this permission notice appear in supporting
 * documentation, and that the name of M.I.T. not be used in
 * advertising or publicity pertaining to distribution of the
 * software without specific, written prior permission.
 * M.I.T. makes no representations about the suitability of
 * this software for any purpose.  It is provided "as is"
 * without express or implied warranty.
 */

#ifndef __EVIL_INET_H__
#define __EVIL_INET_H__


#if ! _WIN32_WINNT >= _WIN32_WINNT_VISTA


/**
 * @file evil_inet.h
 * @brief The file that provides functions ported from Unix in arpa/inet.h.
 * @defgroup Evil_Inet_Group Inet.h functions
 *
 * This header provides functions ported from Unix in arpa/inet.h.
 *
 * @{
 */

/**
 * @brief Convert IPv4 and IPv6 addresses from text to binary form.
 *
 * @param af The address family.
 * @param src The address to convert.
 * @param dst The converted address structure.
 * @return 1 on success, 0 or -1 otherwise.
 *
 * This function converts IPv4 and IPv6 addresses from @p src to the
 * binary form @p dst. The following address families to pass to @p af
 * are currently supported:
 *
 * <ul>
 * <li>i AF_INET: @p src points to a character string containing an IPv4
 * network address in dotted-decimal format, "ddd.ddd.ddd.ddd", where
 * ddd is a decimal number of up to three digits in the range 0 to
 * 255. The address is converted to a struct in_addr and copied to
 * dst, which must be sizeof(struct in_addr) (4) bytes (32 bits) long.
 * <li> AF_INET6: @p     src points to a character string containing an
 * IPv6 network address. The address is converted to a struct in6_addr
 * and copied to dst, which must be sizeof(struct in6_addr) (16) bytes
 * (128 bits) long. The allowed formats for IPv6 addresses follow
 * these rules:
 * <ol>
 * <li>The preferred format is x:x:x:x:x:x:x:x. This form consists of
 * eight hexadecimal numbers, each of which expresses a 16-bit value
 * (i.e., each x can be up to 4 hex digits).
 * <li>A series of contiguous zero values in the preferred format can
 * be abbreviated to ::. Only one instance of :: can occur in an
 * address. For example, the loopback address 0:0:0:0:0:0:0:1 can be
 * abbreviated as ::1. The wildcard address, consisting of all zeros,
 * can be written as ::.
 * <li>An alternate format is useful for expressing IPv4-mapped IPv6
 * addresses. This form is written as x:x:x:x:x:x:d.d.d.d, where the
 * six leading xs are hexadecimal values that define the six
 * most-significant 16-bit pieces of the address (i.e., 96 bits), and
 * the ds express a value in dotted-decimal notation that defines the
 * least significant 32 bits of the address. An example of such an
 * address is :: FFFF:204.152.189.116.
 * </ul>
 * </ul>
 * On success this function returns 1 (network address was successfully
 * converted). 0 is returned if @p src does not contain a character
 * string representing a valid network address in the specified
 * address family. If af does not contain a valid address family, -1
 * is returned and errno is set to EAFNOSUPPORT.
 *
 * @see evil_inet_ntop()
 * @see inet_ntop()
 *
 * Conformity: POSIX.1-2001.
 *
 * Supported OS: Windows XP, CE.
 *
 */
EAPI int evil_inet_pton(int af, const char *src, void *dst);

/**
 * @def inet_pton(x,y,z)
 *
 * Wrapper around evil_inet_pton().
 */
#define inet_pton(x,y,z) evil_inet_pton(x,y,z)

/**
 * @brief Convert IPv4 and IPv6 addresses from binary to text form.
 *
 * @param af The address family.
 * @param src The address structure to convert.
 * @param dst A buffer containing the converted string.
 * @param size The size of the buffer.
 * @return 1 on success, 0 otherwise.
 *
 * This function converts the network address structure @p src in the
 * @p af address family into a character string. The resulting string
 * is copied to the buffer pointed to by @p dst, which must be a
 * non-NULL pointer. The caller specifies the number of bytes
 * available in this buffer in the argument @p size. The following
 * address families to pass to @p af are currently supported:
 *
 * @li AF_INET: @p src points to a struct in_addr (in network byte
 * order) which is converted to an IPv4 network address in the
 * dotted-decimal format, "ddd.ddd.ddd.ddd". The buffer @p dst must be
 * at least INET_ADDRSTRLEN bytes long.
 * @li AF_INET6: @p src points to a struct in6_addr (in network byte
 * order) which is converted to a representation of this address in
 * the most appropriate IPv6 network address format for this
 * address. The buffer @p dst must be at least INET6_ADDRSTRLEN bytes
 * long.
 *
 * On success, this function returns a non-NULL pointer to @p dst. NULL is
 * returned if there was an error, with errno set to indicate the
 * error.
 *
 * @see evil_inet_pton()
 * @see inet_pton()
 *
 * Conformity: POSIX.1-2001.
 *
 * Supported OS: Windows XP, CE.
 *
 */
EAPI const char *evil_inet_ntop(int af, const char *src, void *dst, size_t size);

/**
 * @def inet_ntop(x,y,z,s)
 *
 * Wrapper around evil_inet_ntop().
 */
#define inet_ntop(x,y,z,s) evil_inet_ntop(x,y,z,s)


#endif /* _WIN32_WINNT >= _WIN32_WINNT_VISTA */


/**
 * @}
 */


#endif /* __EVIL_INET_H__ */


src/lib/evil_inet.c

/*
 * Copyright (c) 2004 by Internet Systems Consortium, Inc. ("ISC")
 * Copyright (c) 1996,1999 by Internet Software Consortium.
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND ISC DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS.  IN NO EVENT SHALL ISC BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT
 * OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

/*
 * Modifications: Vincent Torri, for the integration in Evil
 * - modification of the name of some functions
 * * modification of the management of the error
 */

#if ! _WIN32_WINNT >= _WIN32_WINNT_VISTA

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif /* HAVE_CONFIG_H */

#include <stdio.h>
#include <ctype.h>

#ifdef HAVE_ERRNO_H
# include <errno.h>
#endif /* HAVE_ERRNO_H */

#ifndef WIN32_LEAN_AND_MEAN
# define WIN32_LEAN_AND_MEAN
#endif
#include <winsock2.h>
#undef WIN32_LEAN_AND_MEAN

#include "evil_macro.h"
#include "evil_inet.h"
#include "evil_private.h"
#define APICHAR char
#include "evil_print.h"

#ifndef EMSGSIZE
# define EMSGSIZE WSAEMSGSIZE
#endif

#ifndef EAFNOSUPPORT
# define EAFNOSUPPORT WSAEAFNOSUPPORT
#endif

#define SPRINTF(x) ((size_t)sprintf x)

#define ERRNO         ((int)GetLastError())
#define SET_ERRNO(x)  (SetLastError((DWORD)(x)))

#define ISDIGIT(x)  (isdigit((int)  ((unsigned char)x)))
#define ISXDIGIT(x) (isxdigit((int) ((unsigned char)x)))
#define ISUPPER(x)  (isupper((int)  ((unsigned char)x)))

#define NS_IN6ADDRSZ 16
#define NS_INT16SZ 2
#define NS_INADDRSZ sizeof(IN_ADDR)


struct ares_in6_addr {
  union {
    unsigned char _S6_u8[16];
  } _S6_un;
};

const struct ares_in6_addr ares_in6addr_any = { { { 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0 } } };


/*
 * static int
 * inet_net_pton_ipv4(src, dst, size)
 *      convert IPv4 network number from presentation to network format.
 *      accepts hex octets, hex strings, decimal octets, and /CIDR.
 *      "size" is in bytes and describes "dst".
 * return:
 *      number of bits, either imputed classfully or specified with /CIDR,
 *      or -1 if some failure occurred (check errno).  ENOENT means it was
 *      not an IPv4 network specification.
 * note:
 *      network byte order assumed.  this means 192.5.5.240/28 has
 *      0b11110000 in its fourth octet.
 * note:
 *      On Windows we store the error in the thread errno, not
 *      in the winsock error code. This is to avoid loosing the
 *      actual last winsock error. So use macro ERRNO to fetch the
 *      errno this funtion sets when returning (-1), not SOCKERRNO.
 * author:
 *      Paul Vixie (ISC), June 1996
 */
static int
inet_net_pton_ipv4(const char *src, unsigned char *dst, size_t size)
{
  static const char xdigits[] = "0123456789abcdef";
  static const char digits[] = "0123456789";
  int n, ch, tmp = 0, dirty, bits;
  const unsigned char *odst = dst;

  ch = *src++;
  if (ch == '0' && (src[0] == 'x' || src[0] == 'X')
      && ISXDIGIT(src[1])) {
    /* Hexadecimal: Eat nybble string. */
    if (!size)
      goto emsgsize;
    dirty = 0;
    src++;  /* skip x or X. */
    while ((ch = *src++) != '\0' && ISXDIGIT(ch)) {
      if (ISUPPER(ch))
        ch = tolower(ch);
      n = (int)(strchr(xdigits, ch) - xdigits);
      if (dirty == 0)
        tmp = n;
      else
        tmp = (tmp << 4) | n;
      if (++dirty == 2) {
        if (!size--)
          goto emsgsize;
        *dst++ = (unsigned char) tmp;
        dirty = 0;
      }
    }
    if (dirty) {  /* Odd trailing nybble? */
      if (!size--)
        goto emsgsize;
      *dst++ = (unsigned char) (tmp << 4);
    }
  } else if (ISDIGIT(ch)) {
    /* Decimal: eat dotted digit string. */
    for (;;) {
      tmp = 0;
      do {
        n = (int)(strchr(digits, ch) - digits);
        tmp *= 10;
        tmp += n;
        if (tmp > 255)
          goto enoent;
      } while ((ch = *src++) != '\0' &&
               ISDIGIT(ch));
      if (!size--)
        goto emsgsize;
      *dst++ = (unsigned char) tmp;
      if (ch == '\0' || ch == '/')
        break;
      if (ch != '.')
        goto enoent;
      ch = *src++;
      if (!ISDIGIT(ch))
        goto enoent;
    }
  } else
    goto enoent;

  bits = -1;
  if (ch == '/' &&
      ISDIGIT(src[0]) && dst > odst) {
    /* CIDR width specifier.  Nothing can follow it. */
    ch = *src++;    /* Skip over the /. */
    bits = 0;
    do {
      n = (int)(strchr(digits, ch) - digits);
      bits *= 10;
      bits += n;
    } while ((ch = *src++) != '\0' && ISDIGIT(ch));
    if (ch != '\0')
      goto enoent;
    if (bits > 32)
      goto emsgsize;
  }

  /* Firey death and destruction unless we prefetched EOS. */
  if (ch != '\0')
    goto enoent;

  /* If nothing was written to the destination, we found no address. */
  if (dst == odst)
    goto enoent;
  /* If no CIDR spec was given, infer width from net class. */
  if (bits == -1) {
    if (*odst >= 240)       /* Class E */
      bits = 32;
    else if (*odst >= 224)  /* Class D */
      bits = 8;
    else if (*odst >= 192)  /* Class C */
      bits = 24;
    else if (*odst >= 128)  /* Class B */
      bits = 16;
    else                    /* Class A */
      bits = 8;
    /* If imputed mask is narrower than specified octets, widen. */
    if (bits < ((dst - odst) * 8))
      bits = (int)(dst - odst) * 8;
    /*
     * If there are no additional bits specified for a class D
     * address adjust bits to 4.
     */
    if (bits == 8 && *odst == 224)
      bits = 4;
  }
  /* Extend network to cover the actual mask. */
  while (bits > ((dst - odst) * 8)) {
    if (!size--)
      goto emsgsize;
    *dst++ = '\0';
  }
  return (bits);

  enoent:
  SET_ERRNO(ENOENT);
  return (-1);

  emsgsize:
  SET_ERRNO(EMSGSIZE);
  return (-1);
}

static int
getbits(const char *src, int *bitsp)
{
  static const char digits[] = "0123456789";
  int n;
  int val;
  char ch;

  val = 0;
  n = 0;
  while ((ch = *src++) != '\0') {
    const char *pch;

    pch = strchr(digits, ch);
    if (pch != NULL) {
      if (n++ != 0 && val == 0)       /* no leading zeros */
        return (0);
      val *= 10;
      val += (pch - digits);
      if (val > 128)                  /* range */
        return (0);
      continue;
    }
    return (0);
  }
  if (n == 0)
    return (0);
  *bitsp = val;
  return (1);
}

static int
getv4(const char *src, unsigned char *dst, int *bitsp)
{
  static const char digits[] = "0123456789";
  unsigned char *odst = dst;
  int n;
  unsigned int val;
  char ch;

  val = 0;
  n = 0;
  while ((ch = *src++) != '\0') {
    const char *pch;

    pch = strchr(digits, ch);
    if (pch != NULL) {
      if (n++ != 0 && val == 0)       /* no leading zeros */
        return (0);
      val *= 10;
      val += (pch - digits);
      if (val > 255)                  /* range */
        return (0);
      continue;
    }
    if (ch == '.' || ch == '/') {
      if (dst - odst > 3)             /* too many octets? */
        return (0);
      *dst++ = (unsigned char)val;
      if (ch == '/')
        return (getbits(src, bitsp));
      val = 0;
      n = 0;
      continue;
    }
    return (0);
  }
  if (n == 0)
    return (0);
  if (dst - odst > 3)             /* too many octets? */
    return (0);
  *dst++ = (unsigned char)val;
  return (1);
}

static int
inet_net_pton_ipv6(const char *src, unsigned char *dst, size_t size)
{
  static const char xdigits_l[] = "0123456789abcdef",
    xdigits_u[] = "0123456789ABCDEF";
  unsigned char tmp[NS_IN6ADDRSZ], *tp, *endp, *colonp;
  const char *xdigits, *curtok;
  int ch, saw_xdigit;
  unsigned int val;
  int digits;
  int bits;
  size_t bytes;
  int words;
  int ipv4;

  memset((tp = tmp), '\0', NS_IN6ADDRSZ);
  endp = tp + NS_IN6ADDRSZ;
  colonp = NULL;
  /* Leading :: requires some special handling. */
  if (*src == ':')
    if (*++src != ':')
      goto enoent;
  curtok = src;
  saw_xdigit = 0;
  val = 0;
  digits = 0;
  bits = -1;
  ipv4 = 0;
  while ((ch = *src++) != '\0') {
    const char *pch;

    if ((pch = strchr((xdigits = xdigits_l), ch)) == NULL)
      pch = strchr((xdigits = xdigits_u), ch);
    if (pch != NULL) {
      val <<= 4;
      val |= (pch - xdigits);
      if (++digits > 4)
        goto enoent;
      saw_xdigit = 1;
      continue;
    }
    if (ch == ':') {
      curtok = src;
      if (!saw_xdigit) {
        if (colonp)
          goto enoent;
        colonp = tp;
        continue;
      } else if (*src == '\0')
        goto enoent;
      if (tp + NS_INT16SZ > endp)
        return (0);
      *tp++ = (unsigned char)((val >> 8) & 0xff);
      *tp++ = (unsigned char)(val & 0xff);
      saw_xdigit = 0;
      digits = 0;
      val = 0;
      continue;
    }
    if (ch == '.' && ((tp + NS_INADDRSZ) <= endp) &&
        getv4(curtok, tp, &bits) > 0) {
      tp += NS_INADDRSZ;
      saw_xdigit = 0;
      ipv4 = 1;
      break;  /* '\0' was seen by inet_pton4(). */
    }
    if (ch == '/' && getbits(src, &bits) > 0)
      break;
    goto enoent;
  }
  if (saw_xdigit) {
    if (tp + NS_INT16SZ > endp)
      goto enoent;
    *tp++ = (unsigned char)((val >> 8) & 0xff);
    *tp++ = (unsigned char)(val & 0xff);
  }
  if (bits == -1)
    bits = 128;

  words = (bits + 15) / 16;
  if (words < 2)
    words = 2;
  if (ipv4)
    words = 8;
  endp =  tmp + 2 * words;

  if (colonp != NULL) {
    /*
     * Since some memmove()'s erroneously fail to handle
     * overlapping regions, we'll do the shift by hand.
     */
    const ssize_t n = tp - colonp;
    ssize_t i;

    if (tp == endp)
      goto enoent;
    for (i = 1; i <= n; i++) {
      *(endp - i) = *(colonp + n - i);
      *(colonp + n - i) = 0;
    }
    tp = endp;
  }
  if (tp != endp)
    goto enoent;

  bytes = (bits + 7) / 8;
  if (bytes > size)
    goto emsgsize;
  memcpy(dst, tmp, bytes);
  return (bits);

  enoent:
  SET_ERRNO(ENOENT);
  return (-1);

  emsgsize:
  SET_ERRNO(EMSGSIZE);
  return (-1);
}

/*
 * int
 * inet_net_pton(af, src, dst, size)
 *      convert network number from presentation to network format.
 *      accepts hex octets, hex strings, decimal octets, and /CIDR.
 *      "size" is in bytes and describes "dst".
 * return:
 *      number of bits, either imputed classfully or specified with /CIDR,
 *      or -1 if some failure occurred (check errno).  ENOENT means it was
 *      not a valid network specification.
 * note:
 *      On Windows we store the error in the thread errno, not
 *      in the winsock error code. This is to avoid loosing the
 *      actual last winsock error. So use macro ERRNO to fetch the
 *      errno this funtion sets when returning (-1), not SOCKERRNO.
 * author:
 *      Paul Vixie (ISC), June 1996
 */
static int
ares_inet_net_pton(int af, const char *src, void *dst, size_t size)
{
  switch (af) {
  case AF_INET:
    return (inet_net_pton_ipv4(src, dst, size));
  case AF_INET6:
    return (inet_net_pton_ipv6(src, dst, size));
  default:
    SET_ERRNO(EAFNOSUPPORT);
    return (-1);
  }
}

int
evil_inet_pton(int af, const char *src, void *dst)
{
  int result;
  size_t size;

  if (af == AF_INET)
    size = sizeof(struct in_addr);
  else if (af == AF_INET6)
    size = sizeof(struct ares_in6_addr);
  else
  {
    SET_ERRNO(EAFNOSUPPORT);
    return -1;
  }
  result = ares_inet_net_pton(af, src, dst, size);
  if ((result == -1) && (ERRNO == ENOENT))
    return 0;
  return (result > -1 ? 1 : -1);
}

/* const char *
 * inet_ntop4(src, dst, size)
 *     format an IPv4 address, more or less like inet_ntoa()
 * return:
 *     `dst' (as a const)
 * notes:
 *     (1) uses no statics
 *     (2) takes a unsigned char* not an in_addr as input
 * author:
 *     Paul Vixie, 1996.
 */
static const char *
inet_ntop4(const unsigned char *src, char *dst, size_t size)
{
  static const char fmt[] = "%u.%u.%u.%u";
  char tmp[sizeof "255.255.255.255"];

  if (SPRINTF((tmp, fmt, src[0], src[1], src[2], src[3])) > size)
    {
      SET_ERRNO(ENOSPC);
      return (NULL);
    }
    strcpy(dst, tmp);
    return (dst);
}

/* const char *
 * inet_ntop6(src, dst, size)
 *    convert IPv6 binary address into presentation (printable) format
 * author:
 *    Paul Vixie, 1996.
 */
static const char *
inet_ntop6(const unsigned char *src, char *dst, size_t size)
{
  /*
   * Note that int32_t and int16_t need only be "at least" large enough
   * to contain a value of the specified size.  On some systems, like
   * Crays, there is no such thing as an integer variable with 16 bits.
   * Keep this in mind if you think this function should have been coded
   * to use pointer overlays.  All the world's not a VAX.
   */
  char tmp[sizeof("ffff:ffff:ffff:ffff:ffff:ffff:255.255.255.255")];
  char *tp;
  struct {
    long base;
    long len;
  } best, cur;
  unsigned long words[NS_IN6ADDRSZ / NS_INT16SZ];
  int i;

  /*
   * Preprocess:
   *  Copy the input (bytewise) array into a wordwise array.
   *  Find the longest run of 0x00's in src[] for :: shorthanding.
   */
  memset(words, '\0', sizeof(words));
  for (i = 0; i < NS_IN6ADDRSZ; i++)
      words[i / 2] |= (src[i] << ((1 - (i % 2)) << 3));

  best.base = -1;
  cur.base = -1;
  best.len = 0;
  cur.len = 0;

  for (i = 0; i < (NS_IN6ADDRSZ / NS_INT16SZ); i++)
    {
      if (words[i] == 0)
        {
          if (cur.base == -1)
            cur.base = i, cur.len = 1;
          else
            cur.len++;
        }
      else
        {
          if (cur.base != -1)
            {
              if (best.base == -1 || cur.len > best.len)
                best = cur;
              cur.base = -1;
            }
        }
    }
  if (cur.base != -1)
    {
      if (best.base == -1 || cur.len > best.len)
        best = cur;
    }
  if (best.base != -1 && best.len < 2)
    best.base = -1;

  /*
   * Format the result.
   */
  tp = tmp;
  for (i = 0; i < (NS_IN6ADDRSZ / NS_INT16SZ); i++)
    {
      /* Are we inside the best run of 0x00's? */
      if (best.base != -1 && i >= best.base &&
          i < (best.base + best.len))
        {
          if (i == best.base)
             *tp++ = ':';
          continue;
        }
      /* Are we following an initial run of 0x00s or any real hex? */
      if (i != 0)
        *tp++ = ':';
      /* Is this address an encapsulated IPv4? */
      if (i == 6 && best.base == 0 &&
          (best.len == 6 || (best.len == 5 && words[5] == 0xffff)))
        {
          if (!inet_ntop4(src+12, tp, sizeof(tmp) - (tp - tmp)))
            return (NULL);
          tp += strlen(tp);
          break;
        }
        tp += SPRINTF((tp, "%lx", words[i]));
    }

  /* Was it a trailing run of 0x00's? */
  if (best.base != -1 && (best.base + best.len) == (NS_IN6ADDRSZ / NS_INT16SZ))
    *tp++ = ':';
  *tp++ = '\0';

  /*
   * Check for overflow, copy, and we're done.
   */
  if ((size_t)(tp - tmp) > size)
    {
      SET_ERRNO(ENOSPC);
      return (NULL);
    }
  strcpy(dst, tmp);
  return (dst);
}

/* char *
 * inet_ntop(af, src, dst, size)
 *     convert a network format address to presentation format.
 * return:
 *     pointer to presentation format address (`dst'), or NULL (see errno).
 * note:
 *      On Windows we store the error in the thread errno, not
 *      in the winsock error code. This is to avoid loosing the
 *      actual last winsock error. So use macro ERRNO to fetch the
 *      errno this funtion sets when returning NULL, not SOCKERRNO.
 * author:
 *     Paul Vixie, 1996.
 */
static const char *
ares_inet_ntop(int af, const void *src, char *dst, size_t size)
{
  switch (af)
    {
    case AF_INET:
      return (inet_ntop4(src, dst, size));
    case AF_INET6:
      return (inet_ntop6(src, dst, size));
    default:
      SET_ERRNO(EAFNOSUPPORT);
      return (NULL);
    }
  /* NOTREACHED */
}

const char *evil_inet_ntop(int af, const char *src, void *dst, size_t size)
{
  const char *result;
  result = ares_inet_ntop(af, src, dst, size);
  if ((result == NULL) && (ERRNO == ENOSPC))
    return NULL;
  return result;
}

#endif
